django/template/loaders/cached.py

```diff
@@ -16,15 +16,86 @@ class Loader(BaseLoader):
     def __init__(self, engine, loaders):
         self.template_cache = {}
         self.find_template_cache = {}
+        self.get_template_cache = {}
         self.loaders = engine.get_template_loaders(loaders)
         super(Loader, self).__init__(engine)
 
-    def cache_key(self, template_name, template_dirs):
+    def get_contents(self, origin):
+        return origin.loader.get_contents(origin)
+
+    def get_template(self, template_name, template_dirs=None, skip=None):
+        key = self.cache_key(template_name, template_dirs, skip)
+
+        cached = self.get_template_cache.get(key)
+        if cached:
+            if isinstance(cached, TemplateDoesNotExist):
+                raise cached
+            return cached
+
+        try:
+            template = super(Loader, self).get_template(
+                template_name, template_dirs, skip,
+            )
+        except TemplateDoesNotExist as e:
+            self.get_template_cache[key] = e
+            raise e
+        else:
+            self.get_template_cache[key] = template
+
+        return template
+
+    def get_template_sources(self, template_name, template_dirs=None):
+        for loader in self.loaders:
+            for origin in loader.get_template_sources(template_name, template_dirs):
+                yield origin
+
+    def cache_key(self, template_name, template_dirs, skip=None):
+        """
+        Generates a cache key for the template name.
+
+        If skip is provided, cache keys contain the previous matching
+        origin paths. Therefore, given an extends chain like:
+
+            a -> a -> a
+
+        The cache keys will look like:
+
+            1. a
+            2. /path/to/a a
+            3. /otherpath/to/a /path/to/a a
+
+        The cache key only accounts for paths with the same template name. If
+        another extends chain is given, like:
+
+            x -> a -> a -> a
+
+        The keys are generated as:
+
+            1. x
+            2. a
+            3. /path/to/a a
+            4. /otherpath/to/a /path/to/a a
+
+        This means each template object will only be parsed and cached in
+        memory once.
+        """
+        if skip is not None:
+            prefix = ' '.join([
+                origin.name for origin in skip if origin.template_name == template_name
+            ])
+            value = ('%s %s' % (prefix, template_name)).strip()
+        else:
+            value = template_name
+
         if template_dirs:
             # If template directories were specified, use a hash to differentiate
-            return '-'.join([template_name, hashlib.sha1(force_bytes('|'.join(template_dirs))).hexdigest()])
-        else:
-            return template_name
+            return '-'.join([value, hashlib.sha1(force_bytes('|'.join(template_dirs))).hexdigest()])
```

Comment:
`value` could get quite long here. What about a key in the format `<template name>-<hash of skip>-<hash of template_dirs>`?